File: ocaml/bin/p0054.ml
(* Project Euler: Problem 54 *)

(*
  rank:
    0 - High Card: Highest value card.
    1 - One Pair: Two cards of the same value.
    2 - Two Pairs: Two different pairs.
    3 - Three of a Kind: Three cards of the same value.
    4 - Straight: All cards are consecutive values.
    5 - Flush: All cards of the same suit.
    6 - Full House: Three of a kind and a pair.
    7 - Four of a Kind: Four cards of the same value.
    8 - Straight Flush: All cards are consecutive values of same suit.
    9 - Royal Flush: Ten, Jack, Queen, King, Ace, in same suit.

  card:
    Ace(1, only if five-high straight), 2, 3, 4, 5, 6, 7, 8, 9, Ten(10), Jack(11), Queen(12), King(13), Ace(14)

  hand(record):
    (num of suits, [(value, num of value); ...], [(num of value, value); ...])   - lists are descending order
      example: 5C AD 5D AC 9C
                 -> (2, [(14, 2); (5, 2); (9, 1)], [(2, 14); (2, 5); (1, 9)])
             : 9S 3C 9C 5S JS
                 -> (2, [(11, 1); (9, 2); (5, 1); (3, 1)], [(2, 9); (1, 11); (1, 5); (1, 3)])
             : 7H 5D 6S 8H 9H
                 -> (3, [(9, 1); (8, 1); (7, 1); (6, 1); (5, 1)], [(1, 9); (1, 8); (1, 7); (1, 6); (1, 5)])
             : 2D AS 5H 3H 4D
                 -> (3, [(14, 1); (5, 1); (4, 1); (3, 1); (2, 1)], [(1, 14); (1, 5); (1, 4); (1, 3); (1, 2)])
  result(score):
    [rank; weight; ...]]
      example: (2, [(14, 2); (5, 2); (9, 1)], [(2, 14); (2, 5); (1, 9)])
                 -> [2; 14; 5; 9]
               (2, [(11, 1); (9, 2); (5, 1); (3, 1)], [(2, 9); (1, 11); (1, 5); (1, 3)])
                 -> [1; 9; 11; 5; 3]
               (3, [(9, 1); (8, 1); (7, 1); (6, 1); (5, 1)], [(1, 9); (1, 8); (1, 7); (1, 6); (1, 5)])
                 -> [4; 9; 8; 7; 6; 5]
               (3, [(14, 1); (5, 1); (4, 1); (3, 1); (2, 1)], [(1, 14); (1, 5); (1, 4); (1, 3); (1, 2)])
                 -> [4; 5; 4; 3; 2; 1]
 *)

(* ---------------------------------------------------------------- *)

let read_data() =
  let filename = ref "" in
  let anon_fun n = () in
  let speclist = [("-f", Arg.Set_string filename, "<filename>  Set input file name (If not specified, read from stdin)")] in
  Arg.parse speclist anon_fun "Usage:";
  if !filename <> "" then
    let fin = open_in !filename in
    let rec loop acc =
      match input_line fin with
      | l -> loop (l :: acc)
      | exception End_of_file -> close_in fin; List.rev acc
    in
    loop []
  else
    let rec loop acc =
      match read_line () with
      | l -> loop (l :: acc)
      | exception End_of_file -> List.rev acc
    in
    loop []

let make_records hands =
  let make_ninfo nums =
    let card_tbl = Hashtbl.create 5 in
    let rec count_up = function
      | hd :: tl -> if Hashtbl.mem card_tbl hd then (
                      Hashtbl.replace card_tbl hd ((Hashtbl.find card_tbl hd) + 1)
                    ) else (
                      Hashtbl.add card_tbl hd 1
                    );
                    count_up tl
      | [] -> card_tbl
    in
    let c_to_val c =
      List.assoc c [('2', 2); ('3', 3); ('4', 4); ('5', 5); ('6', 6); ('7', 7); ('8', 8);
                    ('9', 9); ('T', 10); ('J', 11); ('Q', 12); ('K', 13); ('A', 14)]
    in
    let nk_card, ck_card = count_up nums
                           |> Hashtbl.to_seq
                           |> Seq.map (fun (c, cnt) -> (((c_to_val c), cnt), (cnt, (c_to_val c))))
                           |> List.of_seq
                           |> List.split in
    let cmp_tpl (m1, n1) (m2, n2) = if m1 <> m2 then m2 - m1 else n2 - n1 in
    List.sort cmp_tpl nk_card, List.sort cmp_tpl ck_card
  in
  let make_sinfo suits =
    let rec aux = function
      | a :: (b :: _ as tl) when a = b -> aux tl
      | a :: b -> a :: aux b
      | [] -> []
    in
    List.length (aux (List.sort compare suits))
  in
  let make_handinfo lst =
    List.map (fun cards -> List.map (fun s -> (s.[0], s.[1])) cards) lst
    |> List.map (List.split)
    |> List.map (fun (nums, suits) -> let vn, nv = make_ninfo nums in (make_sinfo suits, vn, nv))
  in
  make_handinfo hands

let cnvt_hands data =
  List.map (Str.split (Str.regexp " ")) data
  |> List.map (function
         | c1 :: c2 :: c3 :: c4 :: c5 :: tl -> ([c1; c2; c3; c4; c5], tl)
         | _ -> assert false)
  |> List.split

let calc_score (nsuits, vn_lst, nv_lst) =
  let is_straight vn_lst =
    let v, _ = List.split vn_lst in
    match v with
    | v1 :: v2 :: v3 :: v4 :: v5 :: [] when v1 - v2 = 1 && v2 - v3 = 1 && v3 - v4 = 1 && v4 - v5 = 1
      -> Some v
    | v1 :: v2 :: v3 :: v4 :: v5 :: [] when v1 = 14 && v2 = 5 && v3 = 4 && v4 = 3 && v5 = 2
      -> Some [5; 4; 3; 2; 1]      (* five-high straight *)
    | _ -> None
  in
  let other_rank nv_lst =
    let _, v = List.split nv_lst in
    match List.length nv_lst with
    | 2 -> if List.mem_assoc 4 nv_lst then 7 :: v else 6 :: v      (* 7: four of a kind, 6: full house *)
    | 3 -> if List.mem_assoc 3 nv_lst then 3 :: v else 2 :: v      (* 3: three of a kind, 2: two pairs *)
    | 4 -> 1 :: v      (* 1: one pair *)
    | _ -> 0 :: v      (* 0: high card *)
  in
  if nsuits = 1 then
    match is_straight vn_lst with
    | Some v -> if List.hd v = 14 then 9 :: v else 8 :: v      (* 9: royal flush, 8: straight flush *)
    | None -> let v, _ = List.split vn_lst in 5 :: v      (* 5: flush *)
  else
    match is_straight vn_lst with
    | Some v -> 4 :: v      (* 4: straight *)
    | None -> other_rank nv_lst

let judgement recs =
  let rec aux r_lst w1 w2 d =
    match r_lst with
    | (p1, p2) :: tl when p1 = p2 -> aux tl w1 w2 (succ d)      (* draw *)
    | (p1, p2) :: tl when p1 > p2 -> aux tl (succ w1) w2 d      (* player 1 win *)
    | (p1, p2) :: tl when p1 < p2 -> aux tl w1 (succ w2) d      (* player 2 win *)
    | _ -> w1, w2, d
  in
  aux recs 0 0 0

let solve () =
  let p1_hands, p2_hands = read_data () |> cnvt_hands in
  judgement (List.combine (List.map calc_score (make_records p1_hands)) (List.map calc_score (make_records p2_hands)))

let () =
  let p1, p2, d = solve () in
  Printf.printf "Answer:\n  Player1 win = %d\n  Player2 win = %d\n  draw = %d\n" p1 p2 d

File: rust/src/bin/p0054.rs
// Project Euler: Problem 54

/*
  We'll need the following file to run this program.
    - https://projecteuler.net/project/resources/p054_poker.txt

  card rank:
    2, 3, 4, 5, 6, 7, 8, 9, Ten(10), Jack(11), Queen(12), King(13), Ace(14)

  category:
    0 - High Card: Highest value card.
    1 - One Pair: Two cards of the same value.
    2 - Two Pairs: Two different pairs.
    3 - Three of a Kind: Three cards of the same value.
    4 - Straight: All cards are consecutive values.
    5 - Flush: All cards of the same suit.
    6 - Full House: Three of a kind and a pair.
    7 - Four of a Kind: Four cards of the same value.
    8 - Straight Flush: All cards are consecutive values of same suit.
    9 - Royal Flush: Ten, Jack, Queen, King, Ace, in same suit.

  hand info (struct Hand):
    example:
      8H 3D JS 6S 4C -> category: 0, cards: [11, 8, 6, 4, 3]      - HC: High Card [11] @ kicker: [8; 6; 4; 3]
      9S 3C 9C 5S JS -> category: 1, cards: [9, 11, 5, 3]         - OP: One Pair [9] @ kicker : [11; 5; 3]
      5C AD 5D AC 9C -> category: 2, cards: [14, 5; 9]            - TP: Two Pair [14; 5] @ kicker : [9]
      3H 8S 7D 7H 7S -> category: 3, cards: [7, 8, 3]             - TK: Three of a Kind [7] @ kicker : [8; 3]
      7H 5D 6S 8H 9H -> category: 4, cards: [9, 8, 7, 6, 5]       - S:  Straight [9; 8; 7; 6; 5]
      2H 6H 7H QH JH -> category: 5, cards: [12, 11, 7, 6, 2]     - F:  Flush [12; 11; 7; 6; 2]
      4D 8C 8S 4S 4H -> category: 6, cards: [4, 8]                - FH: Full House [4; 8]
      3S 8H 3D 3H 3C -> category: 7, cards: [3, 8]                - FK: Four of a Kind [3] @ kicker : [8]
      8C 6C 7C 5C 9C -> category: 8, cards: [9, 8, 7, 6, 5]       - SF: Straight Flush [9; 8; 7; 6; 5]
      AH JH TH QH KH -> category: 9, cards: [14, 13, 12, 11, 10]  - RF: Royal Flush [14; 13; 12; 11; 10]
*/

use std::cmp::Ordering;
use std::collections::HashMap;

euler::run_solver!(54);

static FILE_DATA: &str = include_str!("../../assets/0054_poker.txt");

#[derive(Ord, PartialOrd, PartialEq, Eq)]
struct Hand {
    category: i64,
    cards: Vec<i64>,
}

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> i64 {
    let all_games = parse_data(data);
    let mut _p1_win: i64 = 0;
    let mut _p2_win: i64 = 0;
    let mut _draw: i64 = 0;
    for cards in all_games {
        let hand_p1 = make_handinfo(&cards[0..5]);
        let hand_p2 = make_handinfo(&cards[5..10]);
        match hand_p1.cmp(&hand_p2) {
            Ordering::Greater => _p1_win += 1,
            Ordering::Less => _p2_win += 1,
            Ordering::Equal => _draw += 1,
        }
    }

    _p1_win
}

fn parse_data(data: &str) -> Vec<Vec<(i64, char)>> {
    let card_num: HashMap<char, i64> = HashMap::from([
        ('2', 2),
        ('3', 3),
        ('4', 4),
        ('5', 5),
        ('6', 6),
        ('7', 7),
        ('8', 8),
        ('9', 9),
        ('T', 10),
        ('J', 11),
        ('Q', 12),
        ('K', 13),
        ('A', 14),
    ]);
    let mut ret: Vec<Vec<(i64, char)>> = Vec::new();

    for line in data.lines() {
        let cards: Vec<(i64, char)> = line
            .split_ascii_whitespace()
            .map(|card| {
                (
                    *card_num.get(&card.chars().next().unwrap()).unwrap(),
                    card.chars().nth(1).unwrap(),
                )
            })
            .collect();
        ret.push(cards);
    }

    ret
}

fn make_handinfo(cards: &[(i64, char)]) -> Hand {
    const CAT_RF: i64 = 9;
    const CAT_SF: i64 = 8;
    const CAT_FK: i64 = 7;
    const CAT_FH: i64 = 6;
    const CAT_F: i64 = 5;
    const CAT_S: i64 = 4;
    const CAT_TK: i64 = 3;
    const CAT_TP: i64 = 2;
    const CAT_OP: i64 = 1;
    const CAT_HC: i64 = 0;

    let (nums, mut suits): (Vec<_>, Vec<_>) = cards.iter().cloned().unzip();

    // Check if a flash or not
    suits.sort();
    suits.dedup();
    let flash = suits.len() == 1;

    // get numbers and their counts
    let mut hand_tmp: Vec<(i64, usize)> = euler::countmap(nums).drain().collect();
    hand_tmp.sort_by(cmp_countmap);
    let (hand_n, hand_c): (Vec<_>, Vec<_>) = hand_tmp.into_iter().unzip();

    let mut rank = Hand {
        category: CAT_HC,
        cards: hand_n,
    };

    if flash {
        if is_straight(&rank.cards) {
            if rank.cards[0] == 14 {
                rank.category = CAT_RF;
            } else {
                rank.category = CAT_SF;
            }
        } else {
            rank.category = CAT_F;
        }
    } else {
        let num_hand_n = rank.cards.len();
        if num_hand_n == 5 {
            if is_straight(&rank.cards) {
                rank.category = CAT_S;
            } else {
                rank.category = CAT_HC;
            }
        } else if num_hand_n == 4 {
            rank.category = CAT_OP;
        } else if num_hand_n == 3 {
            if hand_c[0] == 3 {
                rank.category = CAT_TK;
            } else {
                rank.category = CAT_TP;
            }
        } else if num_hand_n == 2 {
            if hand_c[0] == 4 {
                rank.category = CAT_FK;
            } else {
                rank.category = CAT_FH;
            }
        } else {
            // Not reached
            unreachable!();
        }
    }

    rank
}

fn is_straight(lst: &[i64]) -> bool {
    if lst.len() != 5 {
        false
    } else {
        *lst == (lst[4]..(lst[4] + 5)).rev().collect::<Vec<i64>>()
    }
}

fn cmp_countmap(x: &(i64, usize), y: &(i64, usize)) -> std::cmp::Ordering {
    if x.1 != y.1 {
        (y.1).cmp(&x.1)
    } else {
        (y.0).cmp(&x.0)
    }
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0054() {
        assert_eq!(compute(super::FILE_DATA), 376);
    }
}
